packages/langium/src/documents/document-builder.ts

```diff
@@ -44,65 +40,78 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
-        this.process(document);
-        let diagnostics: Diagnostic[] | undefined = this.connection ? this.validate(document) : undefined;
-        const validator = this.documentValidator;
+    async build(document: LangiumDocument, cancelToken = CancellationToken.None): Promise<BuildResult> {
+        await this.buildDocument(document, cancelToken);
         return {
             parseResult: document.parseResult,
-            get diagnostics() {
-                if (!diagnostics) {
-                    diagnostics = validator.validateDocument(document);
-                }
-                return diagnostics;
-            }
+            diagnostics: await this.validate(document, cancelToken, true)
         };
     }
 
-    validate(document: LangiumDocument): Diagnostic[] | undefined {
-        let diagnostics: Diagnostic[] | undefined;
+    protected async validate(document: LangiumDocument, cancelToken = CancellationToken.None, forceDiagnostics = false): Promise<Diagnostic[]> {
+        let diagnostics: Diagnostic[] = [];
         const validator = this.documentValidator;
-        if (this.connection) {
-            diagnostics = validator.validateDocument(document);
-            // Send the computed diagnostics to VS Code.
-            this.connection.sendDiagnostics({ uri: document.textDocument.uri, diagnostics });
+        if (this.connection || forceDiagnostics) {
+            diagnostics = await validator.validateDocument(document, cancelToken);
+            await interruptAndCheck(cancelToken);
+            if (this.connection) {
+                // Send the computed diagnostics to VS Code.
+                this.connection.sendDiagnostics({ uri: document.textDocument.uri, diagnostics });
+            }
         }
+        await interruptAndCheck(cancelToken);
+        document.state = DocumentState.Validated;
         return diagnostics;
     }
 
-    documentChanged(uri: URI): void {
+    async documentChanged(uri: URI, cancelToken = CancellationToken.None): Promise<void> {
         this.langiumDocuments.invalidateDocument(uri);
         const newDocument = this.langiumDocuments.getOrCreateDocument(uri);
-        this.indexManager.update(newDocument);
-        this.langiumDocuments.all.filter(doc => this.isAffected(doc, uri)).forEach(
-            doc => {
-                this.build(doc);
-                this.indexManager.update(doc);
-            }
-        );
-        this.build(newDocument);
+        await interruptAndCheck(cancelToken);
+        await this.buildDocument(newDocument, cancelToken);
+    }
+
+    protected async buildDocument(document: LangiumDocument, cancelToken: CancellationToken): Promise<void> {
+        const affectedDocuments = this.indexManager.getAffectedDocuments(document);
+        affectedDocuments.forEach(e => {
+            this.linker.unlink(e);
+            e.state = DocumentState.Indexed;
+        });
+        const relevantDocuments = Array.from(new Set([
```

Comment:
What happens with the order of documents after throwing them into a Set? Could this lead to indeterministic execution order? Is there some kind of ordered set (tree set) to avoid this?

Reply:
JavaScript's set use insertion order, they are basically arrays with a `has` method. Same for maps.